### PR TITLE
Add RayTracing example

### PR DESCRIPTION
**Pull request by ttnghia**
image

Here is our first ray-tracing example 😈 
We may need to iterate over the code several more times. Feel free to leave comments.
Also, I'm curious to see how it runs in WebGL.
Enjoy.

### CODE REVIEW

File: src/raytracing/RndGenerators.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <cstdlib>
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Vector2.h>
+#include <Magnum/Math/Vector3.h>
+
+using namespace Magnum;
+namespace Magnum { namespace Examples { namespace Rnd {
+inline Float rand01() {
+    return rand() / static_cast<Float>(RAND_MAX + 1.0f);
```
Comment (mosra) [inline src/raytracing/RndGenerators.h]: (I am by no means an expert on this but) I wonder how much the choice of the random generator contributes to noisiness of the final image. I saw people experimenting with blue noise a lot recently (e.g. [here](https://alain.xyz/blog/raytracing-denoising), [here](http://blog.tuxedolabs.com/2018/12/07/the-importance-of-good-noise.html)). Did the book mention something about this / did you do any experiments in this area?

Using `rand()` is [discouraged](https://codeforces.com/blog/entry/61587), but unfortunately the C++11 RNGs are quite shitty also (see [here](https://arvid.io/2018/06/30/on-cxx-random-number-generator-quality/), [here](https://arvid.io/2018/07/02/better-cxx-prng/)). One of the items on my endless TODO list is adding a sufficiently good and performant random number generator to the Math library, together with the 1D/2D/3D variants, but I think for now switching to `<random>` could be enough.

File: package/ci/appveyor-desktop-mingw.bat [hunk 2]
```diff
@@ -119,6 +119,7 @@ cmake .. ^
     -DWITH_OVR_EXAMPLE=OFF ^
     -DWITH_PICKING_EXAMPLE=ON ^
     -DWITH_PRIMITIVES_EXAMPLE=ON ^
+	-DWITH_RAY_TRACING_EXAMPLE=ON ^
```
Comment (mosra) [inline package/ci/appveyor-desktop-mingw.bat]: Looks like a Tab sneaked in here :D Same in two other files below.

File: src/raytracing/Materials.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+inline bool refract(const Vector3& v, const Vector3& n, Float ni_over_nt, Vector3& refracted) {
+    const Vector3 uv           = v.normalized();
+    const Float   dt           = dot(uv, n);
+    const Float   discriminant = 1.0 - ni_over_nt * ni_over_nt * (1 - dt * dt);
+    if(discriminant > 0) {
+        refracted = ni_over_nt * (uv - n * dt) - n * sqrt(discriminant);
+        return true;
+    }
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/raytracing/Materials.cpp]: Sure. I'll do that.

File: src/raytracing/Objects.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        _center(center), _radiusSqr(radius * radius), _radiusInv(1.0f / radius),
+        _material(_material) {}
+    ~Sphere();
+    bool intersect(const Ray& r, Float t_min, Float t_max, HitInfo& hitInfo) const override;
+private:
+    void computeHitInfo(const Ray& r, Float t, HitInfo& hitInfo) const;
+    Vector3   _center;
+    Float     _radiusSqr;
+    Float     _radiusInv;
+    Material* _material { nullptr }; /* have to use raw pointer as Containers::Pointer cannot be copied */
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/raytracing/Objects.h]: _material must be a shared pointer, but corrade does not support it.

[the remark this answers is not in the document]
Comment (ttnghia) [reply src/raytracing/Objects.h]: During ray tracing, when a ray hits the surface object then the material pointer needs to be copied and return for further shading. Thus, it should be a shared pointer.

File: src/raytracing/RndGenerators.h (hunk 1 above)
Comment (ttnghia) [reply src/raytracing/RndGenerators.h]: Randomness has some effect on the image quality. Yes, I studied this before. Production quality ray tracers implement very complex random generators, like the ones you mentioned. In the extent of this simple example, rand() is good enough.

**Comment by mosra** [comment]
Merged as ecbc805361e872b4ef8cbc06321809e6ceada26a, sorry it took half a century :see_no_evil:

**Pull request closed without merging**